File: backend/app/api/grid.py
"""
GreenCharge — Grid API (P1) — api-contract.md SS3

Routes:
    GET  /api/grid/status
    GET  /api/grid/forecast
    POST /api/grid/signals
    GET  /api/grid/signals
    GET  /api/grid/ev-load
"""

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.config import Settings, get_settings
from app.database import get_db
from app.schemas.grid import (
    GridEVLoadResponse,
    GridForecastResponse,
    GridSignal,
    GridSignalCreate,
    GridSignalsResponse,
    GridStatusResponse,
)
from app.schemas.shared import EnergySlot as EnergySlotSchema
from app.schemas.shared import EVLoad
from app.services.grid.signals import create_signal, list_signals
from app.services.grid.status import build_grid_status, list_forecast_slots
from app.services.shared.ev_load import get_ev_load

router = APIRouter(prefix="/api/grid", tags=["grid"])


@router.get("/status", response_model=GridStatusResponse)
def get_grid_status(
    db: Session = Depends(get_db), settings: Settings = Depends(get_settings)
) -> GridStatusResponse:
    return build_grid_status(db, settings)


@router.get("/forecast", response_model=GridForecastResponse)
def get_grid_forecast(db: Session = Depends(get_db)) -> GridForecastResponse:
    slots = list_forecast_slots(db)
    return GridForecastResponse(
        slots=[EnergySlotSchema.model_validate(s) for s in slots]
    )


@router.post("/signals", response_model=GridSignal, status_code=201)
def post_grid_signal(
    payload: GridSignalCreate, db: Session = Depends(get_db)
) -> GridSignal:
    signal = create_signal(db, payload)
    return GridSignal.model_validate(signal)


@router.get("/signals", response_model=GridSignalsResponse)
def get_grid_signals(db: Session = Depends(get_db)) -> GridSignalsResponse:
    signals = list_signals(db)
    return GridSignalsResponse(signals=[GridSignal.model_validate(s) for s in signals])


@router.get("/ev-load", response_model=GridEVLoadResponse)
def get_grid_ev_load(
    db: Session = Depends(get_db), settings: Settings = Depends(get_settings)
) -> GridEVLoadResponse:
    result = get_ev_load(db, settings)
    return GridEVLoadResponse(
        ev_load=EVLoad(
            current_ev_load_kw=result.current_ev_load_kw,
            scheduled_ev_load_kw=result.scheduled_ev_load_kw,
            peak_ev_load_kw=result.peak_ev_load_kw,
        )
    )

File: frontend/src/services/grid.ts
import type {
  GridForecastResponse,
  GridSignal,
  GridSignalCreate,
  GridSignalsResponse,
  GridStatusResponse,
} from "../types/api";

async function handle<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let message = `Request failed (${res.status})`;
    try {
      const body = (await res.json()) as { detail?: string };
      if (typeof body.detail === "string") message = body.detail;
    } catch {
      // Keep generic HTTP error.
    }
    throw new Error(message);
  }
  return res.json() as Promise<T>;
}

export async function getGridStatus(): Promise<GridStatusResponse> {
  return handle(await fetch("/api/grid/status"));
}

export async function getGridForecast(): Promise<GridForecastResponse> {
  return handle(await fetch("/api/grid/forecast"));
}

export async function getGridSignals(): Promise<GridSignalsResponse> {
  return handle(await fetch("/api/grid/signals"));
}

export async function createGridSignal(payload: GridSignalCreate): Promise<GridSignal> {
  return handle(
    await fetch("/api/grid/signals", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  );
}
